### PR TITLE
Create channel api endpoint

### PR DESCRIPTION
**Pull request by brussee**
Why are channel names required to be unique? -> Design decision.
And why the silent fail? on channel_manager.py:70 -> Solved.

### CODE REVIEW

**Comment by brussee** [comment]
@whirm ready

**Comment by devos50** [comment]
@brussee you should put READY in front of the title of the PR, otherwise @whirm might miss it ;)

File: Tribler/Core/Modules/channel/channel_manager.py [hunk 1]
```diff
@@ -122,3 +124,7 @@ def get_channel_list(self):
         :return: The list of all channel objects.
         """
         return self._channel_list
+
+
+class DuplicateChannelNameError(Exception):
```
Comment (devos50) [inline Tribler/Core/Modules/channel/channel_manager.py]: I would move this to the `exceptions.py` file in `Tribler/Core`.

File: Tribler/Core/Modules/restapi/rest_manager.py [hunk 2]
```diff
@@ -24,10 +27,42 @@ def start(self):
...
+                "message": reason.getErrorMessage()
+            }
+        }
+        if self.site.displayTracebacks:
+            response["error"]["trace"] = format_tb(reason.getTracebackObject())
+
+        body = json.dumps(response)
+        self.setResponseCode(http.INTERNAL_SERVER_ERROR)
+        self.setHeader(b'content-type', self.defaultContentType)
+        self.setHeader(b'content-length', intToBytes(len(body)))
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/rest_manager.py]: Isn't content-length automatically set by the `server.Request` class?

File: Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py [hunk 3]
```diff
@@ -35,18 +41,20 @@ def create_my_channel(self, name, description):
         self.channel_db_handler.on_channel_from_dispersy('fakedispersyid', None, name, description)
         return self.channel_db_handler.getMyChannelId()
 
-    def create_fake_channel(self, channel_name, channel_description):
+    def create_fake_channel(self, name, description, mode=u'closed', rss_url=None):
```
Comment (devos50) [inline Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py]: Why are you not using the `rss_url` parameter in this method?

File: Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py [hunk 4]
```diff
@@ -110,15 +217,15 @@ def test_rss_feeds_endpoint_no_my_channel(self):
 
 
 class TestMyChannelRssEndpoints(AbstractTestMyChannelEndpoints):
-
```
Comment (devos50) [inline Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py]: Very minor detail: I think it's a bit clearer to leave the newlines after a class definition, but please feel free to leave it if you want :)

**Comment by devos50** [comment]
Please also update the documentation with your endpoint description and our new error handling scheme, see [here](https://github.com/Tribler/tribler/blob/devel/doc/Tribler%20REST%20API.md).

**Comment by devos50** [comment]
Other than that, good work 👍

File: Tribler/Core/Modules/restapi/rest_manager.py (hunk 2 above)
Comment (brussee) [inline Tribler/Core/Modules/restapi/rest_manager.py]: The super method that I override here sets this header explicitly.

File: Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py (hunk 3 above)
Comment (brussee) [inline Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py]: I want to have the same method signature as the method in Session that we are replacing.
The same goes for create_fake_channel_with_existing_name

File: Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py (hunk 4 above)
Comment (brussee) [inline Tribler/Test/Core/Modules/RestApi/test_my_channel_endpoints.py]: I'm happy to stick to the current convention.

**Comment by brussee** [comment]
@devos50: Processed all your comments and added documentation.
@whirm: Ready for your review

**Comment by brussee** [comment]
retest this please

**Comment by devos50** [comment]
@brussee I see some files you created in the `android` directory in this PR. Should they be there?

**Comment by brussee** [comment]
@devos50 They are 3 .gitignore files that prevent hundreds of untracked files polluting my git status every time I switch branches from the new Android app to any feature branch from devel.
I don't think they warrant the overhead of a separate pull request and they will end up there eventually.

**Comment by brussee** [comment]
The error callback is not called in the HTTP 500 response tests.

**Comment by brussee** [comment]
Retest this please

**Comment by brussee** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
@whirm Ready for your review

File: Tribler/Core/Modules/restapi/rest_manager.py [hunk 5]
```diff
@@ -24,10 +27,42 @@ def start(self):
...
         Stop the HTTP API and return a deferred that fires when the server has shut down.
         """
         return maybeDeferred(self.site.stopListening)
+
+
+class RESTRequest(server.Request):
+    """
+    This class gracefully takes care of unhandled exceptions raised during the processing of any request.
+    """
+    defaultContentType = b"text/json"
```
Comment (whirm) [inline Tribler/Core/Modules/restapi/rest_manager.py]: @lfdversluis @devos50 after this commit is merged, all the setcontent type calls can be removed from all endpoints

File: Tribler/Core/exceptions.py [hunk 6]
```diff
@@ -48,6 +48,15 @@ def __init__(self, msg=None):
         TriblerException.__init__(self, msg)
 
 
+class DuplicateChannelNameError(TriblerException):
+
+    """ The Channel name already exists in the ChannelManager channel list,
+    i.e., one of your own Channels with the same name already exists. """
+
+    def __init__(self, msg=None):
+        TriblerException.__init__(self, msg)
```
Comment (whirm) [inline Tribler/Core/exceptions.py]: This constructor is doing nothing.

File: Tribler/Test/Core/Modules/RestApi/base_api_test.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def parse_body(self, body):
 
     def parse_response(self, response):
         self.assertEqual(response.code, self.expected_response_code)
-        if response.code == 200 or response.code == 400:
+        if response.code == 200 or response.code == 400 or response.code == 500:
```
Comment (whirm) [inline Tribler/Test/Core/Modules/RestApi/base_api_test.py]:
```python
if response.code in (200,400,500):
```

File: Tribler/Core/exceptions.py (hunk 6 above)
Comment (brussee) [inline Tribler/Core/exceptions.py]: It is a direct copypasta from the existing TriblerExceptions. Do you want me to refactor all exception classes?

**Comment by whirm** [comment]
Also, there's a few things not covered and tests are failing

**Comment by brussee** [comment]
@whirm Those 2 lines that are not covered are mocked to prevent actual channels being created.

The tests that fail are the i2i tests cascading stuff I did not touch.

**Comment by brussee** [comment]
All tests succeeded after the superficial code style change commit.

**Comment by brussee** [comment]
@whirm Ready. 100% test coverage!

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
retest this please

**Comment by devos50** [comment]
@whirm I would like to see this one merged before I'm going to refactor the `mychannel` end points into the `channels` end points.

File: android/TriblerApp/.gitignore [hunk 8]
```diff
@@ -0,0 +1,8 @@
+*.iml
```
Comment (whirm) [inline android/TriblerApp/.gitignore]: Please, merge all the gitignore files with the main one.

**Comment by whirm** [comment]
Looks great, merge the gitignore, rebase and I'll merge

**Comment by brussee** [comment]
@whirm done!

**Pull request merged**